apps/prairielearn/elements/pl-file-upload/pl-file-upload.py

```diff
@@ -86,41 +239,93 @@ def render(element_html: str, data: pl.QuestionData) -> str:
+    opt_file_names = get_file_names_as_array(raw_opt_file_names)
+    raw_file_patterns = pl.get_string_attrib(element, "file-patterns", "")
+    file_patterns = get_file_names_as_array(raw_file_patterns)
+    raw_opt_file_patterns = pl.get_string_attrib(element, "optional-file-patterns", "")
+    opt_file_patterns = get_file_names_as_array(raw_opt_file_patterns)
 
-    # Get submitted answer or return parse_error if it does not exist
-    files = data["submitted_answers"].get(answer_name, None)
-    if not files:
-        add_format_error(data, "No submitted answer for file upload.")
-        return
+    answer_name = get_answer_name(
+        raw_file_names,
+        raw_opt_file_names,
+        raw_file_patterns,
+        raw_opt_file_patterns,
+    )
+
+    # Get submitted answer or return format error if it does not exist
+    files = data["submitted_answers"].get(answer_name)
 
     # We will store the files in the submitted_answer["_files"] key,
-    # so delete the original submitted answer format to avoid
-    # duplication
+    # so delete the original submitted answer format to avoid duplication
     del data["submitted_answers"][answer_name]
 
     try:
-        parsed_files = json.loads(files)
+        parsed_files = json.loads(files)  # pyright: ignore [reportArgumentType]
     except Exception:
-        add_format_error(data, "Could not parse submitted files.")
+        add_format_error(answer_name, data, "Could not parse submitted files.")
         parsed_files = []
 
-    for x in parsed_files:
-        # Filter out any files that were not listed in file_names
-        if x.get("name", "") in required_file_names:
-            pl.add_submitted_file(data, x.get("name", ""), x.get("contents", ""))
-
-    # Validate that all required files are present
     if parsed_files is not None:
-        submitted_file_names = [x.get("name", "") for x in parsed_files]
-        missing_files = [
-            x for x in required_file_names if x not in submitted_file_names
-        ]
+        # Files with HTML characters "<" or ">" should have been rejected client-side already, so we silently filter them here
+        parsed_file_names = [f.get("name", "") for f in parsed_files]
+
+        # Split optional names into two separate lists: wildcard patterns and plain names
+        files_regex = [glob_to_regex(f) for f in file_patterns]
+        opt_files_regex = [glob_to_regex(f) for f in opt_file_patterns]
+
+        # Match submitted and accepted files 1:1 in this order: required_files, then required patterns, then optional files.
+        # All remaining files are then matched 1:n with optional patterns, and eventually discarded in there is no match
+        required_files = [x for x in parsed_file_names if x in file_names]
+        remaining_files = [x for x in parsed_file_names if x not in required_files]
+
+        pattern_files, missing_regex = match_regex_with_files(
+            files_regex,
+            [x for x in remaining_files],
```

Comment:
```suggestion
            remaining_files,
```
I think the `match_regex_with_files` always makes a copy at the start?

Reply:
See other comment, this was to convert into a list, but I'm now just using ``list(...)``.

Reply:
Oh I see. Since the `match_regex_with_files` makes a copy, could that function instead call `list` on the input arguments, then the input type for that function be arbitrary iterables? Not really a big deal but it's nice to avoid making unnecessary copies where possible.

Reply:
See below again. I just had another look and this isn't actually a conversion as I thought it was, so your suggestion is right.